### ISSUE TITLE
`type` declarations should have distinct AST from `var`s

### ISSUE THREAD

**Issue by JamesHenry** [eslint/typescript-eslint-parser#414]
- We currently generate a "special" variable declaration for a type declaration.
- This was done to allow core rules such as `semi` to work with type declarations out of the box.
- However, other core rules (and plugin rules) have issues with this

It leads to issues such as https://github.com/eslint/typescript-eslint-parser/issues/350 https://github.com/eslint/typescript-eslint-parser/issues/359

We should come up with a distinct AST for type declarations and figure out how to make `semi` work in some other way.

**Comment by IanEdington** [comment]
This is throwing `no-restricted-globals`:

```
import * as React from 'react';

import { Name } from './styled';

export interface propTypes {
  name: string;
}

export default class SomeClass extends React.PureComponent<propTypes> {
  render() {
    const { name } = this.props;

    return <Name>{name}</Name>;
  }
}
```

**Comment by majelbstoat** [comment]
```typescript
export default class MyComponent extends React.Component<Props, State> {
  open = () => {
    console.log('something')
  }
}
```

This is also throwing `no-restricted-globals` for `open`.

**Comment by billyvg** [comment]
Is this related to the following throwing `no-eval` errors?
```
export type Spec = {
    eval?: string;
};
```
